File: comind/crates/comind-core/src/types/block_version.rs
use serde::{Deserialize, Serialize};
use sha2::{Sha256, Digest};
use uuid::Uuid;
use crate::{Block, Property, Link};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BlockSnapshot {
    pub block: Block,
    pub properties: Vec<Property>,
    pub relationships: Vec<Link>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BlockVersion {
    pub id: String,
    pub block_id: String,
    pub version: i64,
    pub snapshot: String,
    pub hash: String,
    pub message: Option<String>,
    pub source: String,
    pub restored_from_version_id: Option<String>,
    pub created_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BlockVersionCreateOptions {
    pub block_id: String,
    pub snapshot: String,
    pub hash: String,
    pub message: Option<String>,
    pub source: String,
    pub restored_from_version_id: Option<String>,
}

impl BlockVersion {
    pub fn new(options: BlockVersionCreateOptions, version: i64) -> Self {
        let now = chrono::Utc::now().timestamp_millis();
        BlockVersion {
            id: Uuid::new_v4().to_string(),
            block_id: options.block_id,
            version,
            snapshot: options.snapshot,
            hash: options.hash,
            message: options.message,
            source: options.source,
            restored_from_version_id: options.restored_from_version_id,
            created_at: now,
        }
    }

    pub fn calculate_hash(snapshot: &str) -> String {
        let mut hasher = Sha256::new();
        hasher.update(snapshot.as_bytes());
        let result = hasher.finalize();
        hex::encode(result)
    }
}

File: src/types/blockVersion.ts
import type { Block } from './block'
import type { Property } from '../wasm/types'
import type { Link } from '../wasm/types'

export interface BlockSnapshot {
  block: Block
  properties: Property[]
  relationships: Link[]
}

export interface BlockVersion {
  id: string
  blockId: string
  version: number
  snapshot: string
  hash: string
  message: string | null
  source: 'auto' | 'manual' | 'major_op' | 'app_exit' | 'restore'
  restoredFromVersionId: string | null
  createdAt: number
}

export interface BlockVersionRecord {
  id: string
  blockId: string
  version: number
  snapshot: string
  hash: string
  message: string | null
  source: string
  restoredFromVersionId: string | null
  createdAt: number
}

function sortObjectKeys(obj: Record<string, unknown>): Record<string, unknown> {
  const sorted: Record<string, unknown> = {}
  for (const key of Object.keys(obj).sort()) {
    const value = obj[key]
    if (value !== null && typeof value === 'object' && !Array.isArray(value)) {
      sorted[key] = sortObjectKeys(value as Record<string, unknown>)
    } else {
      sorted[key] = value
    }
  }
  return sorted
}

function sortArrayByKeys<T>(arr: T[]): T[] {
  return arr.map(item => {
    if (typeof item === 'object' && item !== null) {
      return sortObjectKeys(item as Record<string, unknown>) as T
    }
    return item
  }).sort((a, b) => JSON.stringify(a).localeCompare(JSON.stringify(b)))
}

export function serializeSnapshot(snapshot: BlockSnapshot): string {
  const sorted: Record<string, unknown> = {}
  
  sorted['block'] = sortObjectKeys(snapshot.block as unknown as Record<string, unknown>)
  sorted['properties'] = sortArrayByKeys(snapshot.properties as unknown[])
  sorted['relationships'] = sortArrayByKeys(snapshot.relationships as unknown[])
  
  return JSON.stringify(sorted)
}

export async function calculateSnapshotHash(snapshot: BlockSnapshot): Promise<string> {
  const serialized = serializeSnapshot(snapshot)
  const encoder = new TextEncoder()
  const data = encoder.encode(serialized)
  const hashBuffer = await crypto.subtle.digest('SHA-256', data)
  const hashArray = Array.from(new Uint8Array(hashBuffer))
  return hashArray.map(b => b.toString(16).padStart(2, '0')).join('')
}